File: bot/ts/src/utils/logger.ts
import Redis from "ioredis";
import chalk from "chalk";

export enum LogLevel {
    DEBUG = "debug",
    INFO = "info",
    WARN = "warn",
    ERROR = "error"
}

export interface LogEntry {
    timestamp: string;
    level: LogLevel;
    message: string;
    data?: any;
    component?: string;
    userId?: string;
    gameId?: string;
}

export class Logger {
    private redis: Redis | null = null;
    private component: string;

    constructor(component: string = "bot") {
        this.component = component;
        try {
            this.redis = new Redis({
                host: process.env.REDIS_HOST || "localhost",
                port: parseInt(process.env.REDIS_PORT || "6379"),
                maxRetriesPerRequest: 1,
                lazyConnect: true
            });

            // Test connection
            this.redis.ping().catch(() => {
                console.warn(chalk.yellow("Redis connection failed, continuing without Redis logging"));
                this.redis = null;
            });
        } catch (error) {
            console.warn(chalk.yellow("Redis initialization failed, continuing without Redis logging"));
            this.redis = null;
        }
    }

    private async writeLog(level: LogLevel, message: string, data?: any, gameId?: string, userId?: string): Promise<void> {
        const logEntry: LogEntry = {
            timestamp: new Date().toISOString(),
            level,
            message,
            data,
            component: this.component,
            userId,
            gameId
        };

        // Always console log first
        const timestamp = new Date().toLocaleTimeString();
        const prefix = `[${level.toUpperCase()}] ${this.component}:`;

        switch (level) {
            case LogLevel.DEBUG:
                console.log(chalk.gray(`${timestamp} ${prefix} ${message}`), data ? data : "");
                break;
            case LogLevel.INFO:
                console.log(chalk.green(`${timestamp} ${prefix} ${message}`), data ? data : "");
                break;
            case LogLevel.WARN:
                console.log(chalk.yellow(`${timestamp} ${prefix} ${message}`), data ? data : "");
                break;
            case LogLevel.ERROR:
                console.log(chalk.red(`${timestamp} ${prefix} ${message}`), data ? data : "");
                break;
        }

        // Try Redis logging if available
        if (this.redis) {
            try {
                const key = `logs:${this.component}:${Date.now()}`;
                await this.redis.setex(key, 604800, JSON.stringify(logEntry));
                await this.redis.zadd(`logs:${this.component}:index`, Date.now(), key);
            } catch (error) {
                // Disable Redis on error to prevent spam
                console.warn(chalk.yellow("Redis logging disabled due to error"));
                this.redis = null;
            }
        }
    }

    debug(message: string, data?: any, gameId?: string, userId?: string): void {
        this.writeLog(LogLevel.DEBUG, message, data, gameId, userId);
    }

    info(message: string, data?: any, gameId?: string, userId?: string): void {
        this.writeLog(LogLevel.INFO, message, data, gameId, userId);
    }

    warn(message: string, data?: any, gameId?: string, userId?: string): void {
        this.writeLog(LogLevel.WARN, message, data, gameId, userId);
    }

    error(message: string, data?: any, gameId?: string, userId?: string): void {
        this.writeLog(LogLevel.ERROR, message, data, gameId, userId);
    }

    async getLogs(limit: number = 100): Promise<LogEntry[]> {
        if (!this.redis) {
            console.warn("Redis is not available, cannot retrieve logs.");
            return [];
        }
        try {
            const keys = await this.redis.zrevrange(`logs:${this.component}:index`, 0, limit - 1);
            const logs: LogEntry[] = [];

            for (const key of keys) {
                const logData = await this.redis.get(key);
                if (logData) {
                    logs.push(JSON.parse(logData));
                }
            }

            return logs;
        } catch (error) {
            console.error("Failed to retrieve logs from Redis:", error);
            return [];
        }
    }

    async close(): Promise<void> {
        if (this.redis) {
            await this.redis.quit();
        }
    }
}

// Export singleton instance
export const logger = new Logger("poker-bot");

File: bot/api/logger.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/go-redis/redis/v8"
	"golang.org/x/net/context"
)

type LogLevel string

const (
	DEBUG LogLevel = "debug"
	INFO  LogLevel = "info"
	WARN  LogLevel = "warn"
	ERROR LogLevel = "error"
)

type LogEntry struct {
	Timestamp string      `json:"timestamp"`
	Level     LogLevel    `json:"level"`
	Message   string      `json:"message"`
	Data      interface{} `json:"data,omitempty"`
	Component string      `json:"component,omitempty"`
	UserID    string      `json:"userId,omitempty"`
	GameID    string      `json:"gameId,omitempty"`
}

type Logger struct {
	redis        *redis.Client
	component    string
	redisEnabled bool
}

func NewLogger(component string) *Logger {
	redisHost := os.Getenv("REDIS_HOST")
	if redisHost == "" {
		redisHost = "localhost"
	}

	redisPort := os.Getenv("REDIS_PORT")
	if redisPort == "" {
		redisPort = "6379"
	}

	rdb := redis.NewClient(&redis.Options{
		Addr: redisHost + ":" + redisPort,
		DB:   0,
	})

	// Test connection
	ctx := context.Background()
	_, err := rdb.Ping(ctx).Result()
	redisEnabled := err == nil
	if !redisEnabled {
		log.Printf("Redis connection failed, continuing without Redis logging: %v", err)
	}

	return &Logger{
		redis:        rdb,
		component:    component,
		redisEnabled: redisEnabled,
	}
}

func (l *Logger) writeLog(level LogLevel, message string, data interface{}, gameID, userID string) {
	logEntry := LogEntry{
		Timestamp: time.Now().Format(time.RFC3339),
		Level:     level,
		Message:   message,
		Data:      data,
		Component: l.component,
		UserID:    userID,
		GameID:    gameID,
	}

	// Always console log first
	fmt.Printf("[%s] %s: %s\n", string(level), l.component, message)

	// Try Redis logging if enabled
	if l.redisEnabled {
		jsonData, err := json.Marshal(logEntry)
		if err != nil {
			log.Printf("Failed to marshal log entry: %v", err)
			return
		}

		ctx := context.Background()
		key := fmt.Sprintf("logs:%s:%d", l.component, time.Now().UnixNano())

		// Store in Redis with TTL of 7 days
		err = l.redis.SetEX(ctx, key, string(jsonData), 7*24*time.Hour).Err()
		if err != nil {
			log.Printf("Redis logging disabled due to error: %v", err)
			l.redisEnabled = false
			return
		}

		// Add to sorted set for easy querying
		err = l.redis.ZAdd(ctx, fmt.Sprintf("logs:%s:index", l.component), &redis.Z{
			Score:  float64(time.Now().UnixNano()),
			Member: key,
		}).Err()
		if err != nil {
			log.Printf("Failed to add log to index: %v", err)
		}
	}
}

func (l *Logger) Debug(message string, data interface{}, gameID, userID string) {
	l.writeLog(DEBUG, message, data, gameID, userID)
}

func (l *Logger) Info(message string, data interface{}, gameID, userID string) {
	l.writeLog(INFO, message, data, gameID, userID)
}

func (l *Logger) Warn(message string, data interface{}, gameID, userID string) {
	l.writeLog(WARN, message, data, gameID, userID)
}

func (l *Logger) Error(message string, data interface{}, gameID, userID string) {
	l.writeLog(ERROR, message, data, gameID, userID)
}

func (l *Logger) GetLogs(limit int) ([]LogEntry, error) {
	if !l.redisEnabled {
		return []LogEntry{}, nil
	}

	ctx := context.Background()
	indexKey := fmt.Sprintf("logs:%s:index", l.component)

	// Get keys in reverse order (newest first)
	keys, err := l.redis.ZRevRange(ctx, indexKey, 0, int64(limit-1)).Result()
	if err != nil {
		return nil, err
	}

	var logs []LogEntry
	for _, key := range keys {
		logData, err := l.redis.Get(ctx, key).Result()
		if err != nil {
			continue // Skip failed retrievals
		}

		var logEntry LogEntry
		if err := json.Unmarshal([]byte(logData), &logEntry); err != nil {
			continue // Skip malformed entries
		}

		logs = append(logs, logEntry)
	}

	return logs, nil
}

func (l *Logger) Close() error {
	return l.redis.Close()
}
